File: Beecrowd_Parte5/Trilha_Perigosa_2655.java
import java.util.*;

public class Trilha_Perigosa_2655 {
    static int MAX = 100010;
    static int[] road = new int[MAX];

    static int distinctMonsters(int L, int R) {
        Set<Integer> monsters = new HashSet<>();
        for (int i = L; i <= R; i++) {
            monsters.add(road[i]);
        }
        return monsters.size();
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int N = scanner.nextInt();
        int Q = scanner.nextInt();
        int M = scanner.nextInt();
        for (int i = 1; i <= N; i++) {
            road[i] = scanner.nextInt();
        }
        while (Q-- > 0) {
            int type = scanner.nextInt();
            int L = scanner.nextInt();
            int R = scanner.nextInt();
            if (type == 1) {
                System.out.println(distinctMonsters(L, R));
            } else {
                road[L] = R;
            }
        }
    }
}

File: Beecrowd_Parte5/Trilha_Perigosa_2655.c
#include <stdio.h>
#include <stdlib.h>

#define MAX 100010

int road[MAX];

int distinctMonsters(int L, int R) {
    int monsters[51] = {0};
    int count = 0;
    for (int i = L; i <= R; i++) {
        if (monsters[road[i]] == 0) {
            monsters[road[i]] = 1;
            count++;
        }
    }
    return count;
}

int main() {
    int N, Q, M;
    scanf("%d %d %d", &N, &Q, &M);
    for (int i = 1; i <= N; i++) {
        scanf("%d", &road[i]);
    }
    while (Q--) {
        int type, L, R;
        scanf("%d %d %d", &type, &L, &R);
        if (type == 1) {
            printf("%d\n", distinctMonsters(L, R));
        } else {
            road[L] = R;
        }
    }
    return 0;
}